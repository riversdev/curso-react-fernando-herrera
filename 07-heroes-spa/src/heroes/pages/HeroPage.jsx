import { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers'
import 'animate.css'

export const HeroPage = () => {
    const { id } = useParams()
    const navigate = useNavigate()
    const hero = useMemo(() => getHeroById(id), [id])
    const heroImageUrl = `/assets/heroes/${id}.jpg`

    const handleNavigateBack = () => navigate(-1)

    if (!hero)
        return <Navigate to='/marvel' />

    const { superhero, publisher, alter_ego, first_appearance, characters } = hero

    return (
        <div className='row'>
            <div className='col-6'>
                <div className='card'>
                    <div className='row g-0'>
                        <div className='col-md-4'>
                            <img src={heroImageUrl} className='img-fluid rounded-start animate__animated animate__fadeInLeft' alt={superhero} />
                        </div>
                        <div className='col-md-8'>
                            <div className='card-body h-100 d-flex align-items-between flex-wrap'>
                                <div>
                                    <h5 className='card-title'>{superhero}</h5>
                                    <p className='card-text'>{alter_ego}</p>
                                    <p className='card-text'>{publisher}</p>
                                    {alter_ego !== characters && <p className='card-text'>{characters}</p>}
                                    <p className='card-text'>
                                        <small className='text-muted'>{first_appearance}</small>
                                    </p>
                                </div>
                                <div className='w-100 d-flex justify-content-end align-items-end'>
                                    <button
                                        className='btn btn-outline-dark'
                                        onClick={handleNavigateBack}
                                    >
                                        Back
                                    </button>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}