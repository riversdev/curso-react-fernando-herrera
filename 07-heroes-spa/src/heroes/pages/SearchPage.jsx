import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { getHeroesByName } from '../helpers'
import { HeroCard } from '../components'
import 'animate.css'

const useSearch = () => {
    const navigate = useNavigate()
    const location = useLocation()
    const { q = '' } = queryString.parse(location.search)
    const heroes = getHeroesByName(q)
    const { searchText, handleInputChange } = useForm({ searchText: q })

    const handleSubmit = (e) => {
        e.preventDefault()

        navigate(`?q=${searchText}`)
    }

    return {
        q,
        heroes,
        searchText,
        handleInputChange,
        handleSubmit,
    }
}

export const SearchPage = () => {
    const { q, heroes, searchText, handleInputChange, handleSubmit } = useSearch()

    return (
        <>
            <h1>Search</h1>
            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Searching</h4>
                    <hr />
                    <form
                        autoComplete='off'
                        onSubmit={handleSubmit}
                    >
                        <div className='form-floating mb-3'>
                            <input
                                type='text'
                                className='form-control'
                                id='searchText'
                                name='searchText'
                                placeholder='Green Arrow'
                                value={searchText}
                                onChange={handleInputChange}
                            />
                            <label htmlFor='searchText'>Search a hero</label>
                        </div>
                        <button type='submit' className='btn btn-outline-dark'>Submit</button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                            ? <div className='alert alert-info animate__animated animate__fadeIn'>Search a hero</div>
                            : (heroes.length === 0) && <div className='alert alert-danger animate__animated animate__fadeIn'>No hero with <b>{q}</b></div>
                    }
                    <div className='row g-3'>
                        {heroes.map(hero => (
                            <div key={hero.id} className='col-4'>
                                <HeroCard hero={hero} />
                            </div>
                        ))}
                    </div>
                </div>
            </div>
        </>
    )
}