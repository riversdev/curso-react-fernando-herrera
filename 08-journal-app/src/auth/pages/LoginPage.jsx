import { useMemo } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link as RouterLink } from 'react-router-dom'
import { Google } from '@mui/icons-material'
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material'
import { useForm } from '../../hooks'
import { AuthLayout } from '../layout/AuthLayout'
import { startGoogleSignIn, startLoginWithEmailAndPassword } from '../../store/auth'

const initialForm = { email: '', password: '' }

export const LoginPage = () => {
    const { status, errorMessage } = useSelector(state => state.auth)
    const dispatch = useDispatch()
    const { email, password, handleInputChange } = useForm(initialForm)
    const isAuthenticating = useMemo(() => status === 'checking', [status])

    const handleSubmit = e => {
        e.preventDefault()

        if (isAuthenticating) return

        dispatch(startLoginWithEmailAndPassword({ email, password }))
    }

    const handleGoogleSignIn = () => dispatch(startGoogleSignIn())

    return (
        <AuthLayout title='Login'>
            <form onSubmit={handleSubmit}>
                <Grid container>
                    <Grid item xs={12} sx={{ mb: 2 }}>
                        <TextField
                            label='Correo'
                            type='email'
                            placeholder='example@mail.com'
                            fullWidth
                            name='email'
                            value={email}
                            onChange={handleInputChange}
                        />
                    </Grid>
                    <Grid item xs={12} sx={{ mb: 2 }}>
                        <TextField
                            label='Contrasena'
                            type='password'
                            placeholder='Contrasena'
                            fullWidth
                            name='password'
                            value={password}
                            onChange={handleInputChange}
                        />
                    </Grid>
                    <Grid container spacing={2} sx={{ mb: 2 }} >
                        {
                            !!errorMessage &&
                            <Grid item xs={12}>
                                <Alert severity='error'>{errorMessage}</Alert>
                            </Grid>
                        }
                        <Grid item xs={12} sm={6}>
                            <Button
                                type='submit'
                                variant='contained'
                                fullWidth
                                disabled={isAuthenticating}
                            >
                                Login
                            </Button>
                        </Grid>
                        <Grid item xs={12} sm={6}>
                            <Button
                                variant='contained'
                                fullWidth
                                disabled={isAuthenticating}
                                onClick={handleGoogleSignIn}
                            >
                                <Google />
                                <Typography sx={{ ml: 1 }} >Google</Typography>
                            </Button>
                        </Grid>
                    </Grid>
                    <Grid container direction='row' justifyContent='end'>
                        <Link
                            color='inherit'
                            component={RouterLink}
                            to='/auth/register'
                        >
                            Crear una cuenta
                        </Link>
                    </Grid>
                </Grid>
            </form>
        </AuthLayout>
    )
}