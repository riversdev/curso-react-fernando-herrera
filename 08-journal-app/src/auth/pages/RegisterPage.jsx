import { Link as RouterLink } from 'react-router-dom'
import { Button, Grid, Link, TextField, Typography } from '@mui/material'
import { AuthLayout } from '../layout/AuthLayout'
import { useForm } from '../../hooks'
import { useState } from 'react'

const formdata = {
    displayName: '',
    email: '',
    password: '',
}

const formValidations = {
    displayName: [value => value.length >= 1],
    email: [value => value.includes('@'), 'El email debe tener una @ !'],
    password: [value => value.length >= 6, 'El password debe tener almenos 6 caracteres !'],
}

export const RegisterPage = () => {
    const [formSubmitted, setFormSubmitted] = useState(false)
    const {
        formState, displayName, email, password, handleInputChange,
        isFormValid, displayNameValid, emailValid, passwordValid,
    } = useForm(formdata, formValidations)

    const handleSubmit = e => {
        e.preventDefault()

        setFormSubmitted(true)

        if (!isFormValid) return

        console.log(formState)
    }

    return (
        <AuthLayout title='Registro'>
            <form autoComplete='off' onSubmit={handleSubmit}>
                <Grid container>
                    <Grid item xs={12} sx={{ mb: 2 }}>
                        <TextField
                            label='Nombre'
                            type='text'
                            placeholder='Alejandro Rios'
                            fullWidth
                            name='displayName'
                            value={displayName}
                            onChange={handleInputChange}
                            error={!!displayNameValid && formSubmitted}
                            helperText={formSubmitted ? displayNameValid : null}
                        />
                    </Grid>
                    <Grid item xs={12} sx={{ mb: 2 }}>
                        <TextField
                            label='Correo'
                            type='email'
                            placeholder='example@mail.com'
                            fullWidth
                            name='email'
                            value={email}
                            onChange={handleInputChange}
                            error={!!emailValid && formSubmitted}
                            helperText={formSubmitted ? emailValid : null}
                        />
                    </Grid>
                    <Grid item xs={12} sx={{ mb: 2 }}>
                        <TextField
                            label='Contrasena'
                            type='password'
                            placeholder='Contrasena'
                            fullWidth
                            name='password'
                            value={password}
                            onChange={handleInputChange}
                            error={!!passwordValid && formSubmitted}
                            helperText={formSubmitted ? passwordValid : null}
                        />
                    </Grid>
                    <Grid container spacing={2} sx={{ mb: 2 }} >
                        <Grid item xs={12}>
                            <Button
                                type='submit'
                                variant='contained'
                                fullWidth
                            >
                                Crear cuenta
                            </Button>
                        </Grid>
                    </Grid>
                    <Grid container direction='row' justifyContent='end'>
                        <Typography color='gray' sx={{ mr: 2 }}>Ya tienes una cuenta ?</Typography>
                        <Link
                            color='inherit'
                            component={RouterLink}
                            to='/auth/login'
                        >
                            Ingresar
                        </Link>
                    </Grid>
                </Grid>
            </form>
        </AuthLayout>
    )
}