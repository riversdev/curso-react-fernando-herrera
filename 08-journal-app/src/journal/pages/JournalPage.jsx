import { useDispatch, useSelector } from 'react-redux'
import { IconButton } from '@mui/material'
import { JournalLayout } from '../layout/JournalLayout'
import { AddOutlined } from '@mui/icons-material'
import { NoteView, NothingSelectedView } from '../views'
import { startNewNote } from '../../store/journal/thunks'

export const JournalPage = () => {
    const { isSaving, active } = useSelector(state => state.journal)
    const dispatch = useDispatch()

    const handleClickOnNewNote = () => dispatch(startNewNote())

    return (
        <JournalLayout>
            {!!active ? <NoteView /> : <NothingSelectedView />}
            <IconButton
                size='large'
                sx={{
                    color: 'white',
                    backgroundColor: 'error.main',
                    ':hover': {
                        backgroundColor: 'error.main',
                        opacity: 0.9,
                    },
                    ':disabled': {
                        backgroundColor: 'error.main',
                        opacity: 0.7,
                    },
                    position: 'fixed',
                    right: 50,
                    bottom: 50,
                }}
                disabled={isSaving}
                onClick={handleClickOnNewNote}
            >
                <AddOutlined sx={{ fontSize: 30 }} />
            </IconButton>
        </JournalLayout>
    )
}