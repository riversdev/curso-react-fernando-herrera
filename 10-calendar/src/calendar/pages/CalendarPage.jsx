import { useState } from 'react'
import { Calendar } from 'react-big-calendar'
import 'react-big-calendar/lib/css/react-big-calendar.css'
import { calendarLocalizer, getMessagesES } from '../../helpers'
import { CalendarEvent, CalendarModal, FabAddNew, FabDelete, Navbar } from '../components'
import { useCalendarStore, useUiStore } from '../../hooks'

export const CalendarPage = () => {
    const { events, startSetActiveEvent } = useCalendarStore()
    const { startOpenDateModal } = useUiStore()
    const [lastView, setLastView] = useState(localStorage.getItem('lastView') || 'week')

    const eventStyleGetter = (event, start, end, isSelected) => {
        // console.log({ event, start, end, isSelected })

        const style = {
            backgroundColor: '#347CF7',
            borderRadius: '0px',
            opacity: 0.8,
            color: 'white'
        }

        return {
            style
        }
    }

    const handleSelect = (calendarEvent) => startSetActiveEvent({ ...calendarEvent })
    const handleDoubleClick = () => startOpenDateModal()
    const handleViewChange = (view) => {
        setLastView(view)
        localStorage.setItem('lastView', view)
    }

    return (
        <>
            <Navbar />
            <Calendar
                culture='es'
                localizer={calendarLocalizer}
                events={events}
                defaultView={lastView}
                startAccessor='start'
                endAccessor='end'
                style={{ height: 'calc(100vh - 80px)' }}
                messages={getMessagesES()}
                eventPropGetter={eventStyleGetter}
                components={{
                    event: CalendarEvent
                }}
                onSelectEvent={handleSelect}
                onDoubleClickEvent={handleDoubleClick}
                onView={handleViewChange}
            />
            <CalendarModal />
            <FabAddNew />
            <FabDelete />
        </>
    )
}