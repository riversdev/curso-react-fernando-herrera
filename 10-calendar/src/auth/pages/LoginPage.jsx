import { useEffect } from 'react'
import Swal from 'sweetalert2'
import { useAuthStore, useForm } from '../../hooks'
import './LoginPage.css'

const loginForm = { loginEmail: '', loginPassword: '' }
const registerForm = { registerName: '', registerEmail: '', registerPassword: '', registerPassword2: '' }

export const LoginPage = () => {
    const { startLogin, startRegister, errorMessage } = useAuthStore()
    const { loginEmail, loginPassword, handleInputChange: handleLoginInputChange, formState: loginFormState } = useForm(loginForm)
    const { registerName, registerEmail, registerPassword, registerPassword2, handleInputChange: handleRegisterInputChange, formState: registerFormState } = useForm(registerForm)

    const handleLoginSubmit = (e) => {
        e.preventDefault()

        startLogin({ email: loginEmail, password: loginPassword })
    }

    const handleRegisterSubmit = (e) => {
        e.preventDefault()

        if (registerPassword !== registerPassword2) return Swal.fire('Error', 'Las contrasenas no coinciden !', 'error')

        startRegister({ name: registerName, email: registerEmail, password: registerPassword })
    }

    useEffect(() => {
        if (errorMessage) {
            Swal.fire('Error', errorMessage, 'error')
        }
    }, [errorMessage])

    return (
        <div className='container login-container'>
            <div className='row'>
                <div className='col-md-6 login-form-1'>
                    <h3>Ingreso</h3>
                    <form onSubmit={handleLoginSubmit}>
                        <div className='form-group mb-2'>
                            <input
                                type='text'
                                className='form-control'
                                placeholder='Correo'
                                name='loginEmail'
                                value={loginEmail}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='password'
                                className='form-control'
                                placeholder='Contraseña'
                                name='loginPassword'
                                value={loginPassword}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='submit'
                                className='btnSubmit'
                                value='Login'
                            />
                        </div>
                    </form>
                </div>
                <div className='col-md-6 login-form-2'>
                    <h3>Registro</h3>
                    <form onSubmit={handleRegisterSubmit}>
                        <div className='form-group mb-2'>
                            <input
                                type='text'
                                className='form-control'
                                placeholder='Nombre'
                                name='registerName'
                                value={registerName}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='email'
                                className='form-control'
                                placeholder='Correo'
                                name='registerEmail'
                                value={registerEmail}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='password'
                                className='form-control'
                                placeholder='Contraseña'
                                name='registerPassword'
                                value={registerPassword}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='password'
                                className='form-control'
                                placeholder='Repita la contraseña'
                                name='registerPassword2'
                                value={registerPassword2}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className='form-group mb-2'>
                            <input
                                type='submit'
                                className='btnSubmit'
                                value='Crear cuenta'
                            />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}