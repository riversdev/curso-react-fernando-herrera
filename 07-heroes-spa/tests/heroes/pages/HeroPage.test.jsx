import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { HeroPage } from '../../../src/heroes/pages/HeroPage'

const mockUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}))

describe('tests HeroPage', () => {
    beforeEach(() => jest.clearAllMocks())

    test('should display the marvel page if hero is not found', () => {
        const heroIdNotFound = 'dr fate'

        render(
            <MemoryRouter initialEntries={[`/hero/${heroIdNotFound}`]}>
                <Routes>
                    <Route path='/hero/:id' element={<HeroPage />} />
                    {/* Definir la ruta de /marvel porque se navega a ella desde el componente, si no se hace renderiza la pagina de /hero en un bucle */}
                    <Route path='/marvel' element={<h1>This is a Marvel page !</h1>} />
                </Routes>
            </MemoryRouter>
        )

        expect(screen.getByText('This is a Marvel page !')).toBeTruthy()

        // screen.debug()
    })

    test('should display the hero page if the hero exist', () => {
        const heroId = 'dc-arrow'

        render(
            <MemoryRouter initialEntries={[`/hero/${heroId}`]}>
                <Routes>
                    <Route path='/hero/:id' element={<HeroPage />} />
                </Routes>
            </MemoryRouter>
        )

        expect(screen.getByText('Green Arrow')).toBeTruthy()

        // screen.debug()
    })

    test('should navigate at last page when backBtn is clicked', () => {
        const heroId = 'dc-black'

        render(
            <MemoryRouter initialEntries={[`/hero/${heroId}`]}>
                <Routes>
                    <Route path='/hero/:id' element={<HeroPage />} />
                </Routes>
            </MemoryRouter>
        )

        const backBtn = screen.getByRole('button')

        fireEvent.click(backBtn)

        expect(mockUseNavigate).toHaveBeenCalled()
        expect(mockUseNavigate).toHaveBeenCalledWith(-1)

        // screen.debug()
    })

    test('should match with the snapshot', () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/hero/dc-arrow']}>
                <Routes>
                    <Route path='/hero/:id' element={<HeroPage />} />
                </Routes>
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()

        // screen.debug()
    })
})