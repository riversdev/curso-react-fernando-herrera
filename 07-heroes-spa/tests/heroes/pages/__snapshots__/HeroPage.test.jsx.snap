// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests HeroPage should match with the snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-6"
    >
      <div
        class="card"
      >
        <div
          class="row g-0"
        >
          <div
            class="col-md-4"
          >
            <img
              alt="Green Arrow"
              class="img-fluid rounded-start animate__animated animate__fadeInLeft"
              src="/heroes/dc-arrow.jpg"
            />
          </div>
          <div
            class="col-md-8"
          >
            <div
              class="card-body h-100 d-flex align-items-between flex-wrap"
            >
              <div>
                <h5
                  class="card-title"
                >
                  Green Arrow
                </h5>
                <p
                  class="card-text"
                >
                  Oliver Queen
                </p>
                <p
                  class="card-text"
                >
                  DC Comics
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    More Fun Comics #73
                  </small>
                </p>
              </div>
              <div
                class="w-100 d-flex justify-content-end align-items-end"
              >
                <button
                  class="btn btn-outline-dark"
                >
                  Back
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
